// protocol/packet.js
'use strict';

const { handshakeType, contentType } = require('lib/constants');

const {
  CLIENT_HELLO,
  SERVER_HELLO,
  HELLO_VERIFY_REQUEST,
  CERTIFICATE,
  SERVER_KEY_EXCHANGE,
  CERTIFICATE_REQUEST,
  SERVER_HELLO_DONE,
  CERTIFICATE_VERIFY,
  CLIENT_KEY_EXCHANGE,
  FINISHED,
} = handshakeType;

const { CHANGE_CIPHER_SPEC, HANDSHAKE, ALERT, APPLICATION_DATA } = contentType;

/**
 * Create packet sign based on handshake and protocol types.
 * @param {number} protocol
 * @param {number} type
 * @returns {number}
 */
function createSign(protocol, type) {
  return (protocol << 8) | type; // eslint-disable-line no-bitwise
}

/**
 * Get message protocol from state.
 * @param {number} state
 * @returns {number}
 */
function getProtocol(state) {
  return (state >>> 8) & 0xff; // eslint-disable-line no-bitwise
}

/**
 * Get message type from state.
 * @param {number} state
 * @returns {number}
 */
function getType(state) {
  return state & 0xff; // eslint-disable-line no-bitwise
}

const constants = {
  CLIENT_HELLO: createSign(HANDSHAKE, CLIENT_HELLO),
  SERVER_HELLO: createSign(HANDSHAKE, SERVER_HELLO),
  HELLO_VERIFY_REQUEST: createSign(HANDSHAKE, HELLO_VERIFY_REQUEST),
  CERTIFICATE: createSign(HANDSHAKE, CERTIFICATE),
  SERVER_KEY_EXCHANGE: createSign(HANDSHAKE, SERVER_KEY_EXCHANGE),
  CERTIFICATE_REQUEST: createSign(HANDSHAKE, CERTIFICATE_REQUEST),
  SERVER_HELLO_DONE: createSign(HANDSHAKE, SERVER_HELLO_DONE),
  CERTIFICATE_VERIFY: createSign(HANDSHAKE, CERTIFICATE_VERIFY),
  CLIENT_KEY_EXCHANGE: createSign(HANDSHAKE, CLIENT_KEY_EXCHANGE),
  FINISHED: createSign(HANDSHAKE, FINISHED),
  CHANGE_CIPHER_SPEC: createSign(CHANGE_CIPHER_SPEC, 0),
  ALERT: createSign(ALERT, 0),
  APPLICATION_DATA: createSign(APPLICATION_DATA, 0),
  HANDSHAKE: createSign(HANDSHAKE, 0),
};

module.exports = {
  createSign,
  getProtocol,
  getType,
  constants,
};
